v3/src/components/axis/hooks/use-axis.ts

```diff
@@ -17,124 +17,239 @@ export interface IUseAxis {
-
-      select(axisElt)
-        .attr("transform", transform)
-        .transition().duration(duration)
-        .call(axis(scale).tickSizeOuter(0))
-
-      // Onward to gridlines
-      select(axisElt).selectAll('.zero').remove()
-      select(axisElt).selectAll('.grid').remove()
-
-      if (showGridLines) {
-        const tickLength = layout.getAxisLength(otherPlace(axisPlace)) ?? 0,
-          numericScale = scale as ScaleNumericBaseType
+    const axisBounds = layout.getComputedBounds(axisPlace) as AxisBounds,
+      labelBounds = getLabelBounds(label),
+      duration = enableAnimation.current ? transitionDuration : 0,
+      axisIsVertical = isVertical(axisPlace),
+      initialTransform = (place === 'left') ? `translate(${axisBounds.left + axisBounds.width}, ${axisBounds.top})` :
+        `translate(${axisBounds.left}, ${axisBounds.top})`
+
+    const drawAxis = () => {
+        select(axisElt)
+          .attr("transform", initialTransform)
+          .transition().duration(duration)
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          // @ts-ignore types are incompatible
+          .call(axis(scale).tickSizeOuter(0))
+      },
+
+      drawScatterPlotGridLines = () => {
+        select(axisElt).selectAll('.zero, .grid').remove()
+        const tickLength = layout.getAxisLength(otherPlace(axisPlace)) ?? 0
         select(axisElt).append('g')
           .attr('class', 'grid')
           .call(axis(scale).tickSizeInner(-tickLength))
         select(axisElt).select('.grid').selectAll('text').remove()
-
+        const numericScale = scale as ScaleNumericBaseType
         if (between(0, numericScale.domain()[0], numericScale.domain()[1])) {
           select(axisElt).append('g')
             .attr('class', 'zero')
             .call(axis(scale).tickSizeInner(-tickLength).tickValues([0]))
           select(axisElt).select('.zero').selectAll('text').remove()
         }
-      }
+      },
+
+      drawCategoricalAxis = () => {
+        const tickLength = layout.getAxisLength(otherPlace(axisPlace)) ?? 0
+
+        const textHeight = getLabelBounds().height,
+          collision = collisionExists()
+        select(axisElt)
+          .attr("transform", initialTransform)
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          // @ts-ignore types are incompatible
+          .call(axis(scale).tickSizeOuter(0))
+          // Remove everything but the path the forms the axis line
+          .selectAll('g').remove()
+        // select(axisElt).selectAll('line').remove()
+        select(axisElt).append('g')
+          .transition().duration(duration)
+          .attr('transform', `translate(${axisIsVertical ? 0 : bandWidth / 2}, ` +
+            `${axisIsVertical ? bandWidth / 2 : 0})`)
+          .call(axis(scale).tickSizeInner(-tickLength))
+          .selectAll('.domain').remove()
+        let translation = '', rotation = '', textAnchor = 'none'
+        switch (axisPlace) {
```

Comment:
Given the length and complexity of this `drawCategoricalAxis()` function and this `useAxis()` hook, I'd be inclined to move this logic into an external utility function that takes the necessary inputs and returns a structure of the form `{ translation, rotation, textAnchor }`. It could be a custom hook (e.g. `useLabelStyle()`), but it doesn't appear to use any other hooks so it doesn't need to be.

Reply:
I'm not going to try to do this restructuring now before the merge. I certainly agree that the useAxis() hook is ungainly and will consult with you about how best to improve it.